compiler/noirc_evaluator/src/ssa/ir/call_stack.rs

```diff
@@ -0,0 +1,114 @@
+use serde::{Deserialize, Serialize};
+
+use noirc_errors::Location;
+
+pub(crate) type CallStack = im::Vector<Location>;
+
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub(crate) struct CallStackId(u32);
+
+impl CallStackId {
+    pub(crate) fn root() -> Self {
+        Self::new(0)
+    }
+
+    fn new(id: usize) -> Self {
+        Self(id as u32)
+    }
+
+    pub(crate) fn index(&self) -> usize {
+        self.0 as usize
+    }
+
+    pub(crate) fn is_root(&self) -> bool {
+        self.0 == 0
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub(crate) struct LocationNode {
+    pub(crate) parent: Option<CallStackId>,
+    pub(crate) children: Vec<CallStackId>,
+    pub(crate) value: Location,
+}
+
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+pub(crate) struct CallStackHelper {
+    locations: Vec<LocationNode>,
+}
+
+impl CallStackHelper {
+    /// Construct a CallStack from a CallStackId
+    pub(crate) fn get_call_stack(&self, mut call_stack: CallStackId) -> CallStack {
+        let mut result = im::Vector::new();
+        while let Some(parent) = self.locations[call_stack.index()].parent {
+            result.push_back(self.locations[call_stack.index()].value);
+            call_stack = parent;
+        }
+        result
+    }
+
+    /// Returns a new CallStackId which extends the call_stack with the provided call_stack.
+    pub(crate) fn extend_call_stack(
+        &mut self,
+        mut call_stack: CallStackId,
+        locations: &CallStack,
+    ) -> CallStackId {
+        for location in locations {
+            call_stack = self.add_child(call_stack, *location);
+        }
+        call_stack
+    }
+
+    /// Adds a location to the call stack
+    pub(crate) fn add_child(&mut self, call_stack: CallStackId, location: Location) -> CallStackId {
+        if let Some(result) = self.locations[call_stack.index()]
+            .children
+            .iter()
+            .rev()
+            .take(1000)
```

Comment:
I assume we're searching in reverse so that we find recent additions faster; is there any significance as to why give up after 1000 children, after which a duplicate entry is acceptable? Is that value something that was actually observed? 

I wonder if you thought some reverse index from e.g. the hash of the location to the child `CallStackId` could be used to make this faster and easier to reason about? Like `children: HashMap<u64, CallStackId>` where the key is `fxhash::hash64(location)`.